### PR TITLE
Make pip support functional

### PR DESCRIPTION
**Pull request by chmeliik**
CLOUDBLD-11107
CLOUDBLD-11108
CLOUDBLD-12514

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [ ] ~Docs updated (if applicable)~ (will be done in a separate PR)
- [ ] ~Docs links in the code are still valid (if docs were updated)~

### CODE REVIEW

**Comment by chmeliik** [comment]
I kind of accidentally the whole thing

Missing unit tests for fetch_pip_source and documentation (the docs have a follow-up story)

File: tests/unit/test_resolver.py [hunk 1]
```diff
@@ -33,11 +34,18 @@
     environment_variables=[
         {"name": "PIP_INDEX_URL", "value": "file:///some/path", "kind": "literal"},
     ],
+    project_files=[
+        {
+            "abspath": "/your/project/requirements.txt",
+            "template": "I've come to talk with you again.",
```
Comment (brunoapimentel) [inline tests/unit/test_resolver.py:40]: :laughing:

**Comment by brunoapimentel** [comment]
First of all, :rocket: :rocket: :rocket: work here! :rocket: :rocket: :rocket:

I know this is not directly related to this PR, but on my tests here, I felt the checksum warnings are very verbose (made up more than 50% of the log output here).

So maybe change these logs to a debug, or just show a summary instead of logging every single checksum mismatch?
```
2022-12-20 09:51:39,629 WARNING The file yarl-1.8.1.tar.gz has an unexpected checksum value, expected abc06b97407868ef38f3d172762f4069323de52f2b70d133d096a48d72215d28 but computed af887845b8c2e060eb5605ff72b6f2dd2aab7a761379373fd89d314f4752abbf
```

**Comment by brunoapimentel** [comment]
I also feel like we should ease this check here:

https://github.com/containerbuildsystem/cachi2/blob/6d1aab9b55bdad0fd829a9a8abdad9c2cda3b211/cachi2/core/package_managers/pip.py#L110-L117

Neither Cachito or Cachi2 are processable because of this. :laughing: 

Maybe file a ticket for this? (I'm pretty sure there's already one lying around somewhere)

**Comment by chmeliik** [comment]
> I also feel like we should ease this check here:
> 
> https://github.com/containerbuildsystem/cachi2/blob/6d1aab9b55bdad0fd829a9a8abdad9c2cda3b211/cachi2/core/package_managers/pip.py#L110-L117
> 
> Neither Cachito or Cachi2 are processable because of this. laughing
> 
> Maybe file a ticket for this? (I'm pretty sure there's already one lying around somewhere)

Yeah, there is CLOUDBLD-4788

As for processing cachi2, we're also missing pyproject.toml support completely

File: cachi2/core/package_managers/pip.py [hunk 2]
```diff
@@ -58,6 +61,35 @@
 )
 
 
+def fetch_pip_source(request: Request) -> RequestOutput:
```
Comment (brunoapimentel) [inline cachi2/core/package_managers/pip.py:64]: We should probably write a unit test to cover this function.

Comment (chmeliik) [reply cachi2/core/package_managers/pip.py:64]: You were a few minutes faster, done :P

The PR is still a draft, but I think the unit test was one of the last things missing. Other than the `dev` key missing in output.json, which I discovered while writing the unit test.

**Comment by chmeliik** [comment]
> As for processing cachi2, we're also missing pyproject.toml support completely

Filed one for this as well ([CLOUDBLD-12706](https://issues.redhat.com/browse/CLOUDBLD-12706))

File: tests/unit/package_managers/test_pip.py [hunk 3]
```diff
@@ -3318,3 +3320,112 @@ def test_replace_external_requirements(
...
+    output = pip.fetch_pip_source(request)
+
+    expect_package_a = {
+        "name": "foo",
+        "version": "1.0",
+        "type": "pip",
+        "path": Path("."),
+        "dependencies": [
+            # TODO: add 'dev' attribute for non-gomod dependencies
+            {"name": "bar", "version": "https://x.org/bar.zip", "type": "pip"},  # "dev": False},
```
Comment (chmeliik) [inline tests/unit/package_managers/test_pip.py:3394]: I think I'll fix the `dev` key in a follow-up PR. There are at least two ways to do it and I don't want to make this PR too much more complex than it already is

File: cachi2/interface/cli.py [hunk 4]
```diff
@@ -10,7 +10,7 @@
 import typer
 from typer import Argument, Option
```
Comment (brunoapimentel) [inline cachi2/interface/cli.py:11]: Maybe we should remove this line and just use them as `typer.Argument` and `typer.Option`. Or maybe explicitly import everything needed from the typer module.

This way we'll get rid of the CodeQL warning.

**Review by brunoapimentel — APPROVED** [review]
Left a minor comment, but it already LGMT.

File: cachi2/interface/cli.py (hunk 4 above)
Comment (lkolacek) [reply cachi2/interface/cli.py:11]: +1 for `typer.Argument` and `typer.Option` usage

Comment (chmeliik) [reply cachi2/interface/cli.py]: ack, removed the extra imports

**Comment by chmeliik** [comment]
Rebased on main and added `project_files` to int. test data

**Comment by lkolacek** [comment]
Looks great Adam! :+1: Also thanks for updating int. tests due to new changes.

**Review by lkolacek — APPROVED** [review]

**Pull request merged**